server/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java

```diff
@@ -52,14 +54,29 @@
     private final List<SettingUpdater<?>> settingUpdaters = new CopyOnWriteArrayList<>();
     private final Map<String, Setting<?>> complexMatchers;
     private final Map<String, Setting<?>> keySettings;
+    private final Map<Setting<?>, Function<Map.Entry<String, String>, Map.Entry<String, String>>> settingUpgraders;
     private final Setting.Property scope;
     private static final Pattern KEY_PATTERN = Pattern.compile("^(?:[-\\w]+[.])*[-\\w]+$");
     private static final Pattern GROUP_KEY_PATTERN = Pattern.compile("^(?:[-\\w]+[.])+$");
     private static final Pattern AFFIX_KEY_PATTERN = Pattern.compile("^(?:[-\\w]+[.])+[*](?:[.][-\\w]+)+$");
 
-    protected AbstractScopedSettings(Settings settings, Set<Setting<?>> settingsSet, Setting.Property scope) {
+    protected AbstractScopedSettings(
+            final Settings settings,
+            final Set<Setting<?>> settingsSet,
+            final List<SettingUpgrader<?>> settingUpgraders,
```

Comment:
Why not a set, so they are already deduplicated, like with settingsSet above? The order doesn't appear to matter since it is all converted to a map based on the underlying setting anyways?

Reply:
I pushed be2e5de89707ba8770eafd268b1aaa55b82bba50.